### DISCUSSION TITLE
gpiozero simulator on Windows

### DISCUSSION THREAD

**Discussion by Alessandro1985-bo** [gpiozero/gpiozero#965] [General]
Hello,

I am working on a project to simulate a gpiozero on windows by using 2 python project (1 is a normal code to create a project and one is an interrface that are in comunication with server/client). I did some tests with the normal RPi.GPIO code and are working good and now I am creating the gpiozero code.
I did a test with the simply code with led on and led off and is working good, now I would like to create the project with Button, but in the end I have an error because the gpiozero use the RPi.GPIO library.
In web I found the code of gpiozero library and debugging with visual studio, the problem in in the end when link the property of the Button:

```python
TastoUP.when_held=Function1
TastoUP.when_released=Function2
```

I found that is using some code from RPI.GPIO library but I don't where is located on the raspberry in order to open and check, specially the funcions:

```python
def _call_when_changed(self, channel):
        super()._call_when_changed()

def _enable_event_detect(self):
        GPIO.add_event_detect(
           self.number, self._edges,
            callback=self._call_when_changed,
            bouncetime=self._bounce)

def _disable_event_detect(self):
     GPIO.remove_event_detect(self.number)
```

Can someone have an'idea to help me?

Thanks

**Comment by bennuttall** [comment]
Have you read about mock pins? https://gpiozero.readthedocs.io/en/stable/api_pins.html#mock-pins

You don't need to have anything to do with RPi.GPIO - that's a useful pin factory if you're on a Pi.

**Comment by Alessandro1985-bo** [comment]
Thanks for your reply @bennuttall, I am studying what you indicated me, but I don't find the solution. I try to explain better what I would like to do:

by using Raspberry Pi and gpiozero for test the led by the code:

from gpiozero import *
from time import sleep

led = LED(17)

while True:
    led.on()
    sleep(1)
    led.off()
    sleep(1)

By debugging the gpiozero I found that is using the library RPi.GPIO and after some tests I found the solution with my simulation (no raspberry, only windows).


Now I would like to find the solution on my simulator with this simply code that on raspberry works good:

from gpiozero import Button
import time

def ButtonUPPremuto():
    print ("IN")

def ButtonUPRilasciato():
    print ("OUT")

ButtonUP=Button(5, pull_up=False,bounce_time=None,hold_time=0.01)

ButtonUP.when_held=ButtonUPPremuto
ButtonUP.when_released=ButtonUPPRilasciato 

By dedugging the library gpiozero I found that there is an error on the subprocess on rpigpio.py on the functions: 

def _call_when_changed(self, channel):
        super()._call_when_changed()

    def _enable_event_detect(self):
        GPIO.add_event_detect(
            self.number, self._edges,
            callback=self._call_when_changed,
            bouncetime=self._bounce)

    def _disable_event_detect(self):
        GPIO.remove_event_detect(self.number)

that are funcions on RPi.GPIO. For this reason I would like to know if there is a possibility to open the RPi library in order to check what are the functions.

Thanks

**Reply by lurch** [reply]
I'm not sure what you're asking, but if you're looking for the RPi.GPIO sourcecode see https://sourceforge.net/projects/raspberry-gpio-python/

EDIT: But I agree with @bennuttall , it sounds like you ought to be using Mock pins rather than RPi.GPIO pins.
